lib/services/s3.js

```diff
@@ -338,22 +431,172 @@ AWS.util.update(AWS.S3.prototype, {
       404: 'NotFound'
     };
 
+    var req = resp.request;
     var code = resp.httpResponse.statusCode;
     var body = resp.httpResponse.body || '';
+
+    var headers = resp.httpResponse.headers || {};
+    var region = headers['x-amz-bucket-region'] || null;
+    var bucket = req.params.Bucket || null;
+    var bucketRegionCache = req.service.bucketRegionCache;
+    if (region && bucket && region !== bucketRegionCache[bucket]) {
+      bucketRegionCache[bucket] = region;
+    }
+
+    var cachedRegion;
     if (codes[code] && body.length === 0) {
+      if (bucket && !region) {
+        cachedRegion = bucketRegionCache[bucket] || null;
+        if (cachedRegion !== req.httpRequest.region) {
+          region = cachedRegion;
+        }
+      }
       resp.error = AWS.util.error(new Error(), {
-        code: codes[resp.httpResponse.statusCode],
-        message: null
+        code: codes[code],
+        message: null,
+        region: region
       });
     } else {
       var data = new AWS.XML.Parser().parse(body.toString());
+
+      if (data.Region && !region) {
+        region = data.Region;
+        if (bucket && region !== bucketRegionCache[bucket]) {
+          bucketRegionCache[bucket] = region;
+        }
+      } else if (bucket && !region && !data.Region) {
+        cachedRegion = bucketRegionCache[bucket] || null;
+        if (cachedRegion !== req.httpRequest.region) {
+          region = cachedRegion;
+        }
+      }
+
       resp.error = AWS.util.error(new Error(), {
         code: data.Code || code,
         message: data.Message || null,
-        region: data.Region || null
+        region: region
       });
     }
-    resp.request.service.extractRequestIds(resp);
+    req.service.extractRequestIds(resp);
+  },
+
+  /**
+   * If region was not obtained synchronously, then send async request
+   * to get bucket region for errors resulting from wrong region.
+   *
+   * @api private
+   */
+   requestBucketRegion: function requestBucketRegion(resp, done) {
+    var error = resp.error;
+    var req = resp.request;
+    var bucket = req.params.Bucket || null;
+
+    if (!error || !bucket || error.region || req.operation === 'listObjects' ||
+        (AWS.util.isNode() && req.operation === 'headBucket') ||
+        (error.statusCode === 400 && req.operation !== 'headObject') ||
+        regionRedirectErrorCodes.indexOf(error.code) === -1) {
+      return done();
+    }
+
+    var regionReq = req.service.listObjects({Bucket: bucket, MaxKeys: 0});
+    regionReq._requestRegionForBucket = bucket;
+    regionReq.send(function() {
```

Comment:
Is this region available within the callback because the `regionReq` request updates the same cache that the original request uses?